docs/src/proposals/sbf-program-abi-v2.md

```diff
@@ -0,0 +1,103 @@
+We currently only have CBOR in our Cargo.lock, so everything else would add a dependency.
+However, none of these support indirections / references / offsets / pointers and they are all in row-major order (see next point). Therefore, we should design and implement our own extensibile serialization format.
+
+## Nesting order
+Borrowing the terms from table based database systems:
+- Row-major encoding (inner loop over attributes, outer loop over accounts) means first the values of all attributes of one account are encoded, then comes the next account.
+- Column-major encoding (inner loop over accounts, outer loop over attributes) means first the values of one attribute for all accounts are encoded, then comes the next attribute.
+
+Column-major encoding has the following advantages in our use case:
+- Extensibility on an attribute level (but not per account level as that would be wasteful).
+- Read / write access control per attribute (not per account meta data, but the account data is handled separately).
+- Splitting meta data into a local instruction context and a global transaction context.
+
+## Design
+This section describes how the new encoding works in general and how we would use it for the new ABI specifically.
+
+### Encoding
+Everything is in little endian.
+
+```Rust
+struct Map {
+  number_of_entries: u16,
+  attribute: [u16; number_of_entries],
+  value_offset: [u32; number_of_entries],
+  values: [Value],
+}
+
+struct FatPtr {
+  address: u64,
+  length: u32,
+}
+
+enum Value {
+  Direct([u8]),
+  Indirect([FatPtr]),
+}
+```
+
+### Memory Mapping / Regions
+Additional to the mandatory RBPF memory regions (null, program, stack, heap) we would add `2 + number_of_accounts_in_instruction` memory regions to the VM mapping.
+
+The idea is to serialize the meta data of all accounts once per transaction, and only serialize a few attributes like `AccountIsSigner` and `AccountIsWritable` which are currently stored in `KeyedAccounts` per instruction. Then these two meta data regions and the `number_of_accounts_in_instruction` account data regions are mapped directly by pointers and not copied anymore. Furthermore, the `PreAccount`s become obsolete as well because the writability of an account can be enforced by the memory mapping.
+
+#### Read-only meta data region:
+- Transaction context:
+  - `NumberOfAccountsInTransaction = 0`: `u32`
+  - `AccountKey = 1`: `[Pubkey; NumberOfAccountsInTransaction]`
+  - `AccountIsExecutable = 2`: `[bool; NumberOfAccountsInTransaction]`
+  - `AccountOwner = 3`: `[Pubkey; NumberOfAccountsInTransaction]`
+  - `AccountLamports = 4`: `[u64; NumberOfAccountsInTransaction]`
+  - `AccountData = 5`: `[&[u8]; NumberOfAccountsInTransaction]`
+  - `InvocationStackHeight = 6`: `u16`
+  - `InvocationStack = 7`: `[Map; InvocationStackHeight]`
+- Instruction context:
+  - `InstructionData = 8`: `[u8]`
```

Comment:
where is the length of the instruction data communicated?

Reply:
Currently only by the value slice (difference between adjacent `value_offset`s).
However that can might include padding, so yes, we should add an explicit `InstructionDataLength` attribute.